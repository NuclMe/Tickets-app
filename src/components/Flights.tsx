import React, { useEffect } from 'react';
import Tabs from '@mui/material/Tabs';
import Tab from '@mui/material/Tab';
import Box from '@mui/material/Box';
import { useDispatch, useSelector } from 'react-redux';
import { fetchTickets } from '../redux/actions/ticketActions';
import {
  Routes,
  Route,
  Link,
  useLocation,
  useNavigate,
} from 'react-router-dom';
import { Items } from './Items';

export const Flights = () => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchTickets());
  }, [dispatch]);
  const tickets = useSelector((state) => state.ticket.tickets);
  const cheapestTickets = [...tickets].sort((a, b) => a.price - b.price);
  const fastestTickets = [...tickets].sort(
    (a, b) =>
      a.segments[0].duration +
      a.segments[1].duration -
      (b.segments[0].duration + b.segments[1].duration)
  );
  const location = useLocation();
  const navigate = useNavigate();
  const tabIndex = location.pathname === '/fastest' ? 1 : 0;

  const handleChange = (event, newValue) => {
    if (newValue === 0) {
      navigate('/cheapest');
    } else {
      navigate('/fastest');
    }
  };

  return (
    <Box>
      <Tabs
        value={tabIndex}
        onChange={handleChange}
        sx={{
          backgroundColor: 'white',
          borderRadius: '5px',
        }}
      >
        <Tab
          label="САМЫЙ ДЕШЕВЫЙ"
          value={0}
          component={Link}
          to="/cheapest"
          sx={{
            width: '50%',
            maxWidth: '50%',
            borderTop: '1px solid #000',
            borderLeft: '1px solid #000',
            borderBottom: '1px solid #000',
            backgroundColor: tabIndex === 0 ? '#00bfff' : undefined,
            color: tabIndex === 0 ? 'white !important' : undefined,
            borderColor: tabIndex === 0 ? '#00bfff' : undefined,
            boxShadow: 'none',
            transition: 'none',
            '&:hover': { borderColor: '#00bfff' },
            '&:focus': { outline: 'none' },
            '@media (max-width: 768px)': {
              fontSize: '12px',
            },
          }}
        />
        <Tab
          label="САМЫЙ БЫСТРЫЙ"
          value={1}
          component={Link}
          to="/fastest"
          sx={{
            width: '50%',
            maxWidth: '50%',
            borderTop: '1px solid #000',
            borderRight: '1px solid #000',
            borderBottom: '1px solid #000',
            boxShadow: 'none',
            transition: 'none',
            backgroundColor: tabIndex === 1 ? '#00bfff' : undefined,
            color: tabIndex === 1 ? 'white !important' : undefined,
            '&:focus': { outline: 'none' },
            '&:hover': { borderColor: '#00bfff' },
            '@media (max-width: 768px)': {
              fontSize: '12px',
            },
          }}
        />
      </Tabs>
      <Routes>
        <Route path="/cheapest" element={<Items items={cheapestTickets} />} />
        <Route path="/fastest" element={<Items items={fastestTickets} />} />
      </Routes>
    </Box>
  );
};
