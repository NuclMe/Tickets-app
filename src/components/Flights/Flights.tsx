import Box from '@mui/material/Box';
import Tab from '@mui/material/Tab';
import Tabs from '@mui/material/Tabs';
import type { FC } from 'react';
import { Link, useLocation, useNavigate } from 'react-router-dom';

import { device, fontSizes, colors } from '../../styledVars';

export const Flights: FC = () => {
  const location = useLocation();
  const navigate = useNavigate();
  const tabIndex = location.pathname === '/fastest' ? 1 : 0;
  const handleChange = (_: React.SyntheticEvent, newValue: number) => {
    if (newValue === 0) {
      navigate('/cheapest');
    } else {
      navigate('/fastest');
    }
  };

  return (
    <Box>
      <Tabs
        value={tabIndex}
        onChange={handleChange}
        sx={{
          backgroundColor: `${colors.white}`,
          borderRadius: '5px',
        }}
      >
        <Tab
          label="САМЫЙ ДЕШЕВЫЙ"
          value={0}
          component={Link}
          to="/cheapest"
          sx={{
            width: '50%',
            maxWidth: '50%',
            borderTop: `1px solid ${colors.black}`,
            borderLeft: `1px solid ${colors.black}`,
            borderBottom: `1px solid ${colors.black}`,
            backgroundColor: tabIndex === 0 ? `${colors.lightBlue}` : undefined,
            color: tabIndex === 0 ? 'white !important' : undefined,
            borderColor: tabIndex === 0 ? `${colors.lightBlue}` : undefined,
            boxShadow: 'none',
            transition: 'none',
            '&:hover': { borderColor: `${colors.lightBlue}` },
            '&:focus': { outline: 'none' },
            [`${device.tablet}`]: {
              fontSize: fontSizes.small,
            },
          }}
        />
        <Tab
          label="САМЫЙ БЫСТРЫЙ"
          value={1}
          component={Link}
          to="/fastest"
          sx={{
            width: '50%',
            maxWidth: '50%',
            borderTop: `1px solid ${colors.black}`,
            borderRight: `1px solid ${colors.black}`,
            borderBottom: `1px solid ${colors.black}`,
            boxShadow: 'none',
            transition: 'none',
            backgroundColor: tabIndex === 1 ? `${colors.lightBlue}` : undefined,
            color: tabIndex === 1 ? 'white !important' : undefined,
            '&:focus': { outline: 'none' },
            '&:hover': { borderColor: `${colors.lightBlue}` },
            [`${device.tablet}`]: {
              fontSize: fontSizes.small,
            },
          }}
        />
      </Tabs>
    </Box>
  );
};
