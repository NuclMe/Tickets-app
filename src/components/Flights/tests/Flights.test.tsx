import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { describe, it, expect } from 'vitest';

import { Flights } from '../Flights';

const renderWithRouter = (initialEntries = ['/cheapest']) => {
  return render(
    <MemoryRouter initialEntries={initialEntries}>
      <Flights />
    </MemoryRouter>
  );
};

describe('Flights Component', () => {
  it('renders the Flights component via snapshot', () => {
    const { container } = renderWithRouter(['/cheapest']);

    expect(container.firstChild).toMatchSnapshot();
  });
  it('should render both tab buttons', () => {
    renderWithRouter();

    expect(screen.getByText('САМЫЙ ДЕШЕВЫЙ')).toBeInTheDocument();
    expect(screen.getByText('САМЫЙ БЫСТРЫЙ')).toBeInTheDocument();
  });

  it('should have "САМЫЙ ДЕШЕВЫЙ" tab active when on /cheapest route', () => {
    renderWithRouter(['/cheapest']);

    const cheapestTab = screen.getByRole('tab', { name: 'САМЫЙ ДЕШЕВЫЙ' });
    const fastestTab = screen.getByRole('tab', { name: 'САМЫЙ БЫСТРЫЙ' });

    expect(cheapestTab).toHaveAttribute('aria-selected', 'true');
    expect(fastestTab).toHaveAttribute('aria-selected', 'false');
  });

  it('should have "САМЫЙ БЫСТРЫЙ" tab active when on /fastest route', () => {
    renderWithRouter(['/fastest']);

    const cheapestTab = screen.getByRole('tab', { name: 'САМЫЙ ДЕШЕВЫЙ' });
    const fastestTab = screen.getByRole('tab', { name: 'САМЫЙ БЫСТРЫЙ' });

    expect(cheapestTab).toHaveAttribute('aria-selected', 'false');
    expect(fastestTab).toHaveAttribute('aria-selected', 'true');
  });

  it('should navigate to /cheapest when "САМЫЙ ДЕШЕВЫЙ" tab is clicked', () => {
    renderWithRouter(['/fastest']);

    const cheapestTab = screen.getByRole('tab', { name: 'САМЫЙ ДЕШЕВЫЙ' });
    fireEvent.click(cheapestTab);

    expect(cheapestTab).toHaveAttribute('aria-selected', 'true');
  });

  it('should navigate to /fastest when "САМЫЙ БЫСТРЫЙ" tab is clicked', () => {
    renderWithRouter(['/cheapest']);

    const fastestTab = screen.getByRole('tab', { name: 'САМЫЙ БЫСТРЫЙ' });
    fireEvent.click(fastestTab);

    expect(fastestTab).toHaveAttribute('aria-selected', 'true');
  });

  it('should default to cheapest tab when on root route', () => {
    renderWithRouter(['/']);

    const cheapestTab = screen.getByRole('tab', { name: 'САМЫЙ ДЕШЕВЫЙ' });
    const fastestTab = screen.getByRole('tab', { name: 'САМЫЙ БЫСТРЫЙ' });

    expect(cheapestTab).toHaveAttribute('aria-selected', 'true');
    expect(fastestTab).toHaveAttribute('aria-selected', 'false');
  });
});
