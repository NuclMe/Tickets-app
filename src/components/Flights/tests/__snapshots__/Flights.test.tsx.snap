// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Flights Component > renders the Flights component via snapshot 1`] = `
<div
  class="MuiBox-root css-0"
>
  <div
    class="MuiTabs-root css-s6nbqf-MuiTabs-root"
  >
    <div
      class="MuiTabs-scroller MuiTabs-fixed css-s2t35c-MuiTabs-scroller"
      style="overflow: hidden; margin-bottom: 0px;"
    >
      <div
        class="MuiTabs-list MuiTabs-flexContainer css-hzcega-MuiTabs-list"
        role="tablist"
      >
        <a
          aria-selected="true"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-v9sirt-MuiButtonBase-root-MuiTab-root"
          data-discover="true"
          href="/cheapest"
          role="tab"
          tabindex="0"
        >
          САМЫЙ ДЕШЕВЫЙ
        </a>
        <a
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1wut7dv-MuiButtonBase-root-MuiTab-root"
          data-discover="true"
          href="/fastest"
          role="tab"
          tabindex="-1"
        >
          САМЫЙ БЫСТРЫЙ
        </a>
      </div>
      <span
        class="MuiTabs-indicator css-1qltlow-MuiTabs-indicator"
        style="left: 0px; width: 0px;"
      />
    </div>
  </div>
</div>
`;
